### PR TITLE
Fix segment stats in tsdb

### LINKED ISSUE #89609: TSDB: GET _cat/segments failed in time_series index
### Elasticsearch Version

master

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

all

### Problem Description

GET _cat/segments/test_ts?v failed in the time_series index.

### Steps to Reproduce

1. create a more than 2 nodes cluster.
2. create a time_series like:
```
PUT test_ts
{
    "settings": {
      "index.mode": "time_series",
      "index.routing_path": "metricset"
    },
    "mappings": {
      "properties": {
        "metricset": {
          "type": "keyword",
          "time_series_dimension": true
        },
        "k8s": {
          "properties": {
            "pod": {
              "properties": {
                "uid": {
                  "type": "keyword",
                  "time_series_dimension": true
                }
              }
            }
          }
        }
      }
    }
  }
```
3. put some data:
```
POST test_ts/_doc?refresh
{
  "metricset" : "a",
  "@timestamp" : 10
}
```

4. exec `GET _cat/segments/test_ts?v` (to make sure the execute node is not the same with node of the index `test_ts`)
It will cause timeout, and the node will log this:

```
[2022-08-25T11:17:52,569][WARN ][o.e.t.OutboundHandler    ] [node-2] failed to serialize outbound message [Response{620}{false}{false}{false}{class org.elasticsearch.action.support.broadcast.node.TransportBroadcastByNodeAction$NodeResponse}]java.io.IOException: invalid index sort field:<string: "_tsid"> missingValue=SortField.STRING_LAST
	at org.elasticsearch.server@8.5.0-SNAPSHOT/org.elasticsearch.index.engine.Segment.lambda$writeSegmentSort$0(Segment.java:241)
	at org.elasticsearch.server@8.5.0-SNAPSHOT/org.elasticsearch.common.io.stream.StreamOutput.writeArray(StreamOutput.java:916)
	at org.elasticsearch.server@8.5.0-SNAPSHOT/org.elasticsearch.index.engine.Segment.writeSegmentSort(Segment.java:222)
	at org.elasticsearch.server@8.5.0-SNAPSHOT/org.elasticsearch.index.engine.Segment.writeTo(Segment.java:167)
	at org.elasticsearch.server@8.5.0-SNAPSHOT/org.elasticsearch.common.io.stream.StreamOutput.lambda$writeCollection$34(StreamOutput.java:1141)
```

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by nik9000**
The serialization for segment stats was broken for tsdb because we
return a *slightly* different sort configuration. That caused
`_segments` and `_cat/segments` to break when any shard of the tsdb
index is hosted on another node.

Closes #89609

### CODE REVIEW

**Comment by nik9000** [comment]
CC @weizijun

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cat.segments/10_basic.yml [hunk 1]
```diff
@@ -169,3 +169,36 @@
...
+        body:
+          metricset: bar
+          "@timestamp": 2022-01-01T00:00:00Z
+
+  - do:
+      cat.segments: {}
+  - match:
+      $body: |
+        /^(tsdb \s+ 0 \s+ p \s+ \d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3} \s+ _\d (\s\d){3} \s+
+        (\d+|\d+[.]\d+)(kb|b) \s+ \d+ (\s+ (false|true)){2} \s+  \d+\.\d+(\.\d+)? \s+ (false|true) \s? \n?)$/
```
Comment (nik9000) [inline rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cat.segments/10_basic.yml:204]: You can reproduce @weizijun's failure like this:
```
while g -p qa/smoke-test-multinode/ check -Dtests.method='*segments*tsdb*'; do echo ok; done
```

File: server/src/main/java/org/elasticsearch/index/engine/Segment.java [hunk 2]
```diff
@@ -237,6 +244,18 @@ private static void writeSegmentSort(StreamOutput out, Sort sort) throws IOExcep
...
+                if (o.getVersion().before(Version.V_8_5_0)) {
+                    // The closest supported version before 8.5.0 was SortedSet fields, so we mimic that
+                    o.writeByte((byte) 0);
+                    o.writeOptionalBoolean(field.getMissingValue() == null ? null : field.getMissingValue() == SortField.STRING_FIRST);
+                    o.writeBoolean(true);
+                    o.writeBoolean(field.getReverse());
+                } else {
+                    o.writeByte((byte) 5);
+                    o.writeOptionalBoolean(field.getMissingValue() == null ? null : field.getMissingValue() == SortField.STRING_FIRST);
+                    o.writeBoolean(field.getReverse());
```
Comment (csoulios) [inline server/src/main/java/org/elasticsearch/index/engine/Segment.java:257]: Not sure if I understand this part very well. Don't we need to write a boolean for the selector? 

In all other cases (`type = 0`) I see one byte and thre booleans are written. Not sure why `type = 5` requires only 2 booleans.

File: server/src/main/java/org/elasticsearch/index/engine/Segment.java [hunk 3]
```diff
@@ -189,6 +189,13 @@ private static Sort readSegmentSort(StreamInput in) throws IOException {
                 if (missingFirst != null) {
                     fields[i].setMissingValue(missingFirst ? SortedSetSortField.STRING_FIRST : SortedSetSortField.STRING_LAST);
                 }
+            } else if (type == 5) {
```
Comment (csoulios) [inline server/src/main/java/org/elasticsearch/index/engine/Segment.java:192]: Where can I see what `type = 5` is? A comment would help here, a constant would be even better, I think

**Review by csoulios — COMMENTED** [review]
I am not familiar with the `Segment` class and its inner workings.

I have left couple comments/questions, but I leave it up to the other reviewers to approve this PR.

**Review by jpountz — APPROVED** [review]
The change makes sense to me, `_tsid` is our first field that we index as a single-value field which can be used for index sorting.

File: server/src/main/java/org/elasticsearch/index/engine/Segment.java (hunk 2 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/engine/Segment.java:257]: If we have a selector we use the other case above. In this case we just don't have one.

**Comment by nik9000** [comment]
> The change makes sense to me, _tsid is our first field that we index as a single-value field which can be used for index sorting.

@jpountz could you comment on why we aren't using Lucene's serialization for these? Or why we don't make our own object rather than this brittle thing? I don't particularly want to change now, but it seems like we wouldn't try to serialize with the `instanceof` stuff if we were building it now.

File: server/src/main/java/org/elasticsearch/index/engine/Segment.java (hunk 3 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/engine/Segment.java:192]: Yeah.  I was following the old way and it didn't have them. I didn't want to rock the boat too much here, but there isn't any reason not to try and at least make it more readable. I'll add some constants.

**Comment by jpountz** [comment]
I would guess that one reason is because segment stats added the sorting configuration to its output before Lucene exposed serialization for SortFields, which used to be a codec implementation before. @romseygeek might have more info about whether moving to Lucene serialization would be possible and a good idea. FWIW I see we have custom serialization logic for `SortField` in `org.elasticsearch.common.lucene.Lucene#readSortField` as well.

**Comment by nik9000** [comment]
> prg.elasticsearch.common.lucene.Lucene#readSortField

Oh no! It's totally different!

**Pull request merged**